snuba/replacer.py

```diff
@@ -381,3 +417,85 @@ def flush_batch(self, batch: Sequence[Replacement]) -> None:
             logger.info(
                 "Optimized %s partitions on %s" % (num_dropped, clickhouse_read.host)
             )
+
+    def _message_already_processed(self, metadata: ReplacementMessageMetadata) -> bool:
+        """
+        Figure out whether or not the message was already processed.
+
+        Check if there exists a recorded offset that took too long to execute.
+        If there is, we can conclude whether or not to process the incoming
+        message by comparing the incoming message to the recorded offset.
+        """
+        key = self._build_topic_group_index_key(metadata)
+        self._reset_offset_check(key)
+
+        # float("-inf") implies this is the first message this worker is processing since startup
+        if self.__prolonged_offset == float("-inf"):
+            processed_offset = redis_client.get(key)
+            try:
+                self.__prolonged_offset = (
+                    -1 if processed_offset is None else int(processed_offset)
+                )
```

Comment:
I don't think this will work as you want.
Different partitions of the same topic can have reached very different offsets. Each partitions has a monotonic offset, these offsets are independent from each other. 
So on partition 1 the last offset may be 5000 while on partition 2 the last offset can be 200,000. 
This means you will have to keep track the last offset from all partitions, you cannto have a single `self.__prolonged offset`.

Let's say you have an issue when committing partition 2 (offset 200,000).
Let's say you load the offset from redis for partition 2: that yields 200,000. Now you load that in memory, from redis when you restart. If you get a message from partition 1 with offset 5001 you will discard it while you never processed it.
Basically you cannot compare offsets across partitions. You can only compare offsets of the same partition.